### PR TITLE
[broker][monitoring] add per-subscription EntryFilter metrics

### PR DESCRIPTION
**Pull request by tjiuming**
Fixes https://github.com/apache/pulsar/issues/16596

### Motivation

add per-subscription `EntryFilter` metrics for PersistentTopic and NonPersistentTopic.

### Modifications

Add the following metrics to Prometheus:
1. pulsar_subscription_filter_processed_msg_count
2. pulsar_subscription_filter_accepted_msg_count
3. pulsar_subscription_filter_rejected_msg_count
4. pulsar_subscription_filter_rescheduled_msg_count

Add the following metrics to admin-api:
1. filterProcessedMsgCount
2. filterAcceptedMsgCount
3. filterRejectedMsgCount
4. filterRescheduledMsgCount

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Comment by tjiuming** [comment]
@eolivelli @michaeljmarshall PTAL

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java [hunk 1]
```diff
@@ -748,6 +749,14 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
                 topicStatsStream.writePair("msgThroughputOut", subMsgThroughputOut);
                 topicStatsStream.writePair("msgRateRedeliver", subMsgRateRedeliver);
                 topicStatsStream.writePair("type", subscription.getTypeString());
+
+                // Write entry filter stats
+                Dispatcher dispatcher0 = subscription.getDispatcher();
+                topicStatsStream.writePair("throughEntryFilterMsgs", dispatcher0.getThroughFilterMsgCount());
+                topicStatsStream.writePair("entryFilterAccepted", dispatcher0.getFilterAcceptedMsgCount());
```
Comment (asafm) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java:756]: Just occurred to me: A subscription has a list of Entry Filter. Maybe it's more correct: `entryFiltersAccepted`.
Also, on the first metric you end with units `Msgs`, but all other 3 no. I would `entryFiltersAcceptedMsgs` and same for other 2 metrics below

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java [hunk 2]
```diff
@@ -281,6 +281,16 @@ static void printTopicStats(SimpleTextOutputStream stream, String cluster, Strin
                     subsStats.msgDropRate, splitTopicAndPartitionIndexLabel);
             metric(stream, cluster, namespace, topic, n, "pulsar_subscription_consumers_count",
                     subsStats.consumersCount, splitTopicAndPartitionIndexLabel);
+
+            metric(stream, cluster, namespace, topic, n, "pulsar_subscription_through_filter_msg_count",
```
Comment (asafm) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java:285]: Same comment above, but here it should to be consistent `pulsar_subscription_filter_processed_msgs_count`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/SubscriptionStatsTest.java [hunk 3]
```diff
@@ -134,4 +144,123 @@ public void testNonContiguousDeletedMessagesRanges() throws Exception {
...
+            producer.newMessage().property("ACCEPT", " ").value(UUID.randomUUID().toString()).send();
+        }
+        for (int i = 0; i < 100; i++) {
+            producer.newMessage().property("REJECT", " ").value(UUID.randomUUID().toString()).send();
+        }
+        for (int i = 0; i < 100; i++) {
+            producer.newMessage().property("RESCHEDULE", " ").value(UUID.randomUUID().toString()).send();
+        }
+
+        for (;;) {
```
Comment (asafm) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/SubscriptionStatsTest.java:195]: `while (true) {`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/SubscriptionStatsTest.java [hunk 4]
```diff
@@ -134,4 +144,123 @@ public void testNonContiguousDeletedMessagesRanges() throws Exception {
...
+                    .filter(m -> m.tags.get("subscription").equals(subName) && m.tags.get("topic").equals(topic))
+                    .mapToDouble(m-> m.value).sum();
+
+            Assert.assertEquals(filterAccepted, 100);
+            if (isPersistent) {
+                Assert.assertEquals(filterRejected, 100);
+                Assert.assertEquals(throughFilter, filterAccepted + filterRejected + filterRescheduled, 0.01 * throughFilter);
+            }
+        } else {
+            Assert.assertEquals(throughFilterMetrics.size(), 0);
```
Comment (asafm) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/SubscriptionStatsTest.java:242]: Why 0?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java [hunk 5]
```diff
@@ -51,6 +52,10 @@ public abstract class AbstractBaseDispatcher extends EntryFilterSupport implemen
 
     protected final ServiceConfiguration serviceConfig;
     protected final boolean dispatchThrottlingOnBatchMessageEnabled;
+    private final LongAdder throughFilterMsgs = new LongAdder();
+    private final LongAdder filterAcceptedMsgs = new LongAdder();
```
[the remark this answers is not in the document]
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java:56]: the origin issue requires messages metrics

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/SubscriptionStatsTest.java (hunk 4 above)
Comment (tjiuming) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/SubscriptionStatsTest.java:242]: per-subscription metrics are disabled when `exposeTopicLevelMetricsInPrometheus=false`, so it has to be 0

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -1756,6 +1756,13 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
                 topicStatsStream.writePair("totalNonContiguousDeletedMessagesRange",
                         subscription.getTotalNonContiguousDeletedMessagesRange());
                 topicStatsStream.writePair("type", subscription.getTypeString());
+
+                Dispatcher dispatcher0 = subscription.getDispatcher();
+                topicStatsStream.writePair("throughEntryFilterMsgs", dispatcher0.getThroughFilterMsgCount());
+                topicStatsStream.writePair("entryFilterAccepted", dispatcher0.getFilterAcceptedMsgCount());
+                topicStatsStream.writePair("entryFilterRejected", dispatcher0.getFilterRejectedMsgCount());
+                topicStatsStream.writePair("entryFilterRescheduled", dispatcher0.getFilterRescheduledMsgCount());
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1764]: I know this is being discussed elsewhere in the PR, but just want to point out that the methods have `Msg` and the metrics have `entry`. We should make sure that the end result is consistent.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaeljmarshall) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java:56]: This is a good discussion to have. Note that the `filterEntriesForConsumer` method, which we are instrumenting uses `entries` throughout. At first, I thought entries would make more sense. However, I think total messages makes sense here. One interesting point is that the `msgOutCounter` measures the total messages by summing all of the messages within a batched message. Ultimately, I think the final solution should be consistent.

The current solution looks good to me.

**Review by michaeljmarshall — COMMENTED** [review]
Looks good. I think there might need to be some minor clean up before we can merge it though.

Thank you @tjiuming for working on this!

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java [hunk 7]
```diff
@@ -748,6 +749,14 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
                 topicStatsStream.writePair("msgThroughputOut", subMsgThroughputOut);
                 topicStatsStream.writePair("msgRateRedeliver", subMsgRateRedeliver);
                 topicStatsStream.writePair("type", subscription.getTypeString());
+
+                // Write entry filter stats
+                Dispatcher dispatcher0 = subscription.getDispatcher();
```
[the remark this answers is not in the document]
Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java:754]: you're right, dispatcher can be null

**Review by asafm — APPROVED** [review]

**Review by michaeljmarshall — COMMENTED** [review]
@tjiuming - all of the logic looks correct to me. I have some minor comments about method and variable names that I think we should update or discuss before merging this. Let me know what you think. Thanks!

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 8]
```diff
@@ -1756,6 +1756,19 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
                 topicStatsStream.writePair("totalNonContiguousDeletedMessagesRange",
                         subscription.getTotalNonContiguousDeletedMessagesRange());
                 topicStatsStream.writePair("type", subscription.getTypeString());
+
+                Dispatcher dispatcher0 = subscription.getDispatcher();
+                if (null != dispatcher0) {
+                    topicStatsStream.writePair("entryFilterProccessedMsgs",
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1762]:
```suggestion
                    topicStatsStream.writePair("entryFilterProcessedMsgs",
```

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/SubscriptionStatsImpl.java [hunk 9]
```diff
@@ -137,6 +137,14 @@ public class SubscriptionStatsImpl implements SubscriptionStats {
     /** SubscriptionProperties (key/value strings) associated with this subscribe. */
     public Map<String, String> subscriptionProperties;
 
+    public long throughFilterMsgCount;
```
Comment (michaeljmarshall) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/SubscriptionStatsImpl.java:140]:
```suggestion
    public long filterProcessedMsgCount;
```

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/SubscriptionStats.java [hunk 10]
```diff
@@ -126,4 +126,12 @@ public interface SubscriptionStats {
 
     /** The serialized size of non-contiguous deleted messages ranges. */
     int getNonContiguousDeletedMessagesRangesSerializedSize();
+
+    long getThroughFilterMsgCount();
```
Comment (michaeljmarshall) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/SubscriptionStats.java:130]:
```suggestion
    long getFilterProcessedMsgCount();
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java [hunk 11]
```diff
@@ -748,6 +749,20 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
...
+                if (null != dispatcher0) {
+                    topicStatsStream.writePair("entryFilterProcessedMsgs",
+                            dispatcher0.getFilterProcessedMsgCount());
+                    topicStatsStream.writePair("entryFilterAcceptedMsgs",
+                            dispatcher0.getFilterAcceptedMsgCount());
+                    topicStatsStream.writePair("entryFilterRejectedMsgs",
+                            dispatcher0.getFilterRejectedMsgCount());
+                    topicStatsStream.writePair("entryFilterRescheduledMsgs",
+                            dispatcher0.getFilterRescheduledMsgCount());
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java:763]: We use `MsgCount` elsewhere, and it'd be great to make it consistent. Also, I don't think we need the `entry` filter prefix.  
```suggestion
                    topicStatsStream.writePair("filterProcessedMsgCount",
                            dispatcher0.getFilterProcessedMsgCount());
                    topicStatsStream.writePair("filterAcceptedMsgCount",
                            dispatcher0.getFilterAcceptedMsgCount());
                    topicStatsStream.writePair("filterRejectedMsgCount",
                            dispatcher0.getFilterRejectedMsgCount());
                    topicStatsStream.writePair("filterRescheduledMsgCount",
                            dispatcher0.getFilterRescheduledMsgCount());
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 12]
```diff
@@ -1756,6 +1756,19 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
...
+                Dispatcher dispatcher0 = subscription.getDispatcher();
+                if (null != dispatcher0) {
+                    topicStatsStream.writePair("entryFilterProcessedMsgs",
+                            dispatcher0.getFilterProcessedMsgCount());
+                    topicStatsStream.writePair("entryFilterAcceptedMsgs",
+                            dispatcher0.getFilterAcceptedMsgCount());
+                    topicStatsStream.writePair("entryFilterRejectedMsgs",
+                            dispatcher0.getFilterRejectedMsgCount());
+                    topicStatsStream.writePair("entryFilterRescheduledMsgs",
+                            dispatcher0.getFilterRescheduledMsgCount());
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1769]: Same suggestion.
```suggestion
                    topicStatsStream.writePair("filterProcessedMsgCount",
                            dispatcher0.getFilterProcessedMsgCount());
                    topicStatsStream.writePair("filterAcceptedMsgCount",
                            dispatcher0.getFilterAcceptedMsgCount());
                    topicStatsStream.writePair("filterRejectedMsgCount",
                            dispatcher0.getFilterRejectedMsgCount());
                    topicStatsStream.writePair("filterRescheduledMsgCount",
                            dispatcher0.getFilterRescheduledMsgCount());
```

**Review by michaeljmarshall — COMMENTED** [review]
@tjiuming - I have one last request for changes before we merge this. I want to make sure the stat names exposed via the admin api match the prometheus metric names as closely as possible. Thanks again for working on this!

**Comment by tjiuming** [comment]
@michaeljmarshall many thanks for you review! I've resolved all the suggestions, PTAL

**Comment by michaeljmarshall** [comment]
@tjiuming - thanks! Do you feel strongly about keeping `Msgs` in the admin stats output, or can we change it to `MsgCount`? I like the latter because it matches the prometheus metric exactly then.

**Comment by tjiuming** [comment]
> @tjiuming - thanks! Do you feel strongly about keeping `Msgs` in the admin stats output, or can we change it to `MsgCount`? I like the latter because it matches the prometheus metric exactly then.

it makes sense, and I've resolved them

**Review by michaeljmarshall — APPROVED** [review]

**Comment by michaeljmarshall** [comment]
Thanks @tjiuming! Great work!

**Pull request merged**